src/ivectorbin/ivector-extract-dense.cc

```diff
@@ -80,107 +140,142 @@ int main(int argc, char *argv[]) {
+        if (!posterior_reader.HasKey(utt)) {
+          KALDI_WARN << "No posteriors for utterance " << utt;
+          num_err++;
+          continue;
+        }
+        const Matrix<BaseFloat> &mat = feature_reader.Value(utt);
+        Posterior posterior = posterior_reader.Value(utt);
       
-      if (static_cast<int32>(posterior.size()) != mat.NumRows()) {
-        KALDI_WARN << "Size mismatch between posterior " << posterior.size()
-                   << " and features " << mat.NumRows() << " for utterance "
-                   << utt;
-        num_err++;
-        continue;
-      }
+        if (static_cast<int32>(posterior.size()) != mat.NumRows()) {
+          KALDI_WARN << "Size mismatch between posterior " << posterior.size()
+                     << " and features " << mat.NumRows() << " for utterance "
+                     << utt;
+          num_err++;
+          continue;
+        }
+
+        double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
 
-      double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
+        double this_t = opts.acoustic_weight * TotalPosterior(posterior),
+            max_count_scale = 1.0;
+        if (opts.max_count > 0 && this_t > opts.max_count) {
+          max_count_scale = opts.max_count / this_t;
+          KALDI_LOG << "Scaling stats for utterance " << utt << " by scale "
+                    << max_count_scale << " due to --max-count="
+                    << opts.max_count;
+          this_t = opts.max_count;
+        }
+        ScalePosterior(opts.acoustic_weight * max_count_scale,
+                       &posterior);
+        // note: now, this_t == sum of posteriors.
 
-      double this_t = opts.acoustic_weight * TotalPosterior(posterior),
-          max_count_scale = 1.0;
-      if (opts.max_count > 0 && this_t > opts.max_count) {
-        max_count_scale = opts.max_count / this_t;
-        KALDI_LOG << "Scaling stats for utterance " << utt << " by scale "
-                  << max_count_scale << " due to --max-count="
-                  << opts.max_count;
-        this_t = opts.max_count;
+        int32 num_chunks = std::max((int32)ceil((mat.NumRows() - chunk_size + period) / static_cast<BaseFloat>(period)), 1);
+        Matrix<BaseFloat> ivectors(num_chunks, extractor.IvectorDim());
+        for (int32 i = 0; i < num_chunks; i++) {
+          Vector<BaseFloat> ivector(extractor.IvectorDim());
+          int32 window = std::min(chunk_size, mat.NumRows() - i * period);
+          SubMatrix<BaseFloat> sub_mat(mat, i * period, window, 0, mat.NumCols());
+          IvectorExtract(extractor, utt, Matrix<BaseFloat>(sub_mat),
+  		         std::vector<std::vector<std::pair<int32, BaseFloat> > >
+  		         (&posterior[i * period], &posterior[i * period + window]),
+      		         &ivector, auxf_ptr);
+          ivectors.CopyRowFromVec(ivector, i);
+
+	  std::stringstream ss;
+	  double start = iter->second[seg_num]->start;
```

Comment:
seg_num is really a string, right? This name makes me think it's an int. It might be better to rename this to seg_id or simply seg.